contracts/BookMarketplace.sol

```diff
@@ -44,15 +48,98 @@ contract BookMarketplace is ERC721URIStorage {
+        bool allowBid,
+        Bid[] bidList,
+        bool sold
+    );
+
+    /* modifiers */
+    
+    
+    // Modifier prevents function from being called by anyone other than the market owner
+    modifier onlyMarketOwner() {
+        require(
+            marketOwner == msg.sender,
+            "Only the marketplace owner can call this function."
+        );
+        _;
+    }
+
+    /* constructor */
+
     constructor() ERC721("BookExchange Tokens", "BOOK") {
-        owner = payable(msg.sender);
+        marketOwner = payable(msg.sender);
+    }
+
+    /* functions */
+
+    function getMarketOwnerAddress() public view returns(address) {
+        return marketOwner;
+    }
+
+    // change the price of a listing
+    function updateInstantPrice(uint256 tokenId, uint256 newPrice) public payable onlyMarketOwner {
+        Listing storage listing = idToMarketItem[tokenId];
+        listing.instantPrice = newPrice;
+        idToMarketItem[tokenId] = listing;
+    }
+
+    // get the price of a listing
+    function getInstantPrice(uint256 tokenId) public view returns (uint256) {
+        Listing storage listing = idToMarketItem[tokenId];
+        return listing.instantPrice;
+    }
+
+    // Adds bids to a bid list by tokenId
+    function addBid(uint256 tokenId, uint256 bidAmount) public payable {
+        Listing storage listing = idToMarketItem[tokenId];
+        require(
+            listing.allowBid == true,
+            "This item does not allow bids."
+        );
+        require(
+            listing.sold == false,
+            "This item has been sold."
+        );
+        require(
+            bidAmount >= listing.instantPrice,
+            "Bid must be greater than or equal to the current price."
+        );
+        require(
+            msg.value == bidAmount,
+            "Bid amount must be equal to the amount sent." 
+        );
+        Bid memory newBid;
+        newBid.bidder = msg.sender;
+        newBid.bidAmount = bidAmount;
+        listing.bidList.push(newBid);
+        idToMarketItem[tokenId] = listing;
+    }
+
+    // Returns the bid list for a given tokenId
+    function getBidList(uint256 tokenId) public view returns (Bid[] memory) {
+        Listing storage listing = idToMarketItem[tokenId];
+        return listing.bidList;
+    }
+
+    // marks a listing as sold
+    function markListingAsSold(uint256 tokenId) public onlyMarketOwner {
+        Listing storage listing = idToMarketItem[tokenId];
+        listing.sold = true;
+        idToMarketItem[tokenId] = listing;
     }
 
     /* Updates the listing price of the contract */
-    function updateListingPrice(uint256 _listingPrice) public payable {
+    function updateListingPrice(uint256 _listingPrice) public payable onlyMarketOwner {
         require(
-            owner == msg.sender,
-            "Only marketplace owner can update listing price."
+            _listingPrice > 0,
+            "Listing price must be greater than 0."
```

Comment:
This is an odd case so I need your opinions. What If we have a promo week where the listing price is free? @MichaelFauht feels free to give your 2 cents.

Reply:
Id be fine with making it >=. I was mostly thinking about the case of a negative values. Ill wait for @MichaelFauht to chime in.

Reply:
>= 0 sounds good to me. Having promotions like that is a great idea @axelnguyen0701